자동차 경주 5단계: 자동차 경주(리팩토링)

jdalma opened the pull request:
안녕하세요 ㅎㅎ 마지막 5단계도 잘 부탁드립니다.

1. CarMove라는 애매한 이름보다는 이름 길이 검증 책임을 추가하면서 RacingRule로 변경하였습니다.
   - 하지만 Controller에서는 이름 길이 검증을 하고, Stadium에서 숫자 생성기를 사용하게 되어 서로 다른 계층에서 같은 객체를 사용하는 찝찝함이 있습니다.
2. RacingController를 추가하니 View 작업과 완전히 분리된 것 같습니다.
   - 말씀해주신 "경기 결과" 책임을 가진 객체를 추가하니 수월했던 것 같습니다.
3. 전진할 수 있는지 판단하는 클래스 안에서 상수들을 관리했는데 상수들을 주입할 수 있도록 RacingProperty로 밖에서 주입할 수 있도록 변경했습니다.
   - 테스트하기에는 더 수월해졌지만 괜찮은 판단인지는 모르겠네요..
4. 외부에서 주입을 많이 하다보니 테스트 코드를 작성할 때 setup이 길어지는 느낌이 들어서.. 리뷰어님이 보시기에는 어떤지 궁금하네요

jdalma committed:
refactor: 4단계 리뷰 내용 적용

1. RacingHistories에서 RacingResult로 네이밍 수정함
2. RacingResult에서 Winner를 구하도록 수정함
3. RandomGenerator를 NumberGenerator로 네이밍 수정 및 Number 타입만 가능하도록 수정함
4. View에서 결과 내용을 출력하는 메서드 정리

jdalma committed:
feat: 자동차 이름 검증 추가

jdalma committed:
refactor: Controller 레이어 추가

1. RacingController를 추가하여 View와 자동차 경주 도메인을 완전히 분리
2. CarMove 객체를 RacingRule로 변경하여 이름 길이 검증 책임 추가
3. RacingRule에서 자동차 전진 판단과 정수 랜덤 생성을 상수로 관리하였는데 외부에서 주입하기 위해 RacingProperty로 추출

jdalma committed:
test: RacingController 테스트 코드 추가

jdalma committed:
style: ktlint 통과 못하는 부분 수정

namjackson commented on src/main/kotlin/racingcar/RacingController.kt:
```diff
@@ -0,0 +1,25 @@
+) {
+
+    fun start(rule: RacingRule): RacingResult {
+        val names = nameOfCars.split(",")
+        require(rule.nameValidate(names)) {
+            Message.CAR_NAME_LENGTH_EXCEPTION.message
+        }
```
RacingController 생성시점에 검증해도 좋을거같아요! 
init 블록을 활용해보면 어떨까요?

namjackson commented on src/main/kotlin/racingcar/domain/NameValidation.kt:
```diff
@@ -0,0 +1,9 @@
+package racingcar.domain
+
+fun interface NameValidation {
+    fun test(names: List<String>): Boolean
```
test라는 네이밍보단 check같은 네이밍이 조금더 어울리진않을까요?

namjackson commented on src/main/kotlin/racingcar/domain/RacingRule.kt:
```diff
@@ -0,0 +1,10 @@
+package racingcar.domain
+
+class RacingRule(
```
RacingRule이라는 객체의 책임은 무엇일까요?
여러개의 책임 ( NumberGenerator , NameValidation, Property)이 혼합된 형태는 아닐까요?
한번 추가적으로 고민해봐도 좋을거 같아요

namjackson replied:
> 하지만 Controller에서는 이름 길이 검증을 하고, Stadium에서 숫자 생성기를 사용하게 되어 서로 다른 계층에서 같은 객체를 사용하는 찝찝함이 있습니다.

말씀해주신 대로 굳이 세가지의 역할을 하나로 묶을 이유는 없을거 같아요!
RacingStadium에서는 NumberGenerator가
Controller에서는 NameValidation만 필요하진 않을까요?

namjackson commented on src/main/kotlin/racingcar/domain/RacingResult.kt:
```diff
@@ -0,0 +1,20 @@
+package racingcar.domain
+
+class RacingResult {
```
테스트코드를 작성해보면 어떨까요?
생성시점에 주입받는 구조를 작성해보면 어떨까요?
`class RacingResult( val allRounds: List<List<RacingHistory>>)`

namjackson commented on src/main/kotlin/racingcar/domain/RacingRule.kt:
```diff
@@ -0,0 +1,10 @@
+package racingcar.domain
+
+class RacingRule(
+    private val random: NumberGenerator<Int>,
+    private val validation: NameValidation,
+    private val property: RacingProperty
```
`외부에서 주입을 많이 하다보니 테스트 코드를 작성할 때 setup이 길어지는 느낌이 들어서.. 리뷰어님이 보시기에는 어떤지 궁금하네요`
위 코멘트 처럼, 불필요한 곳에서도 주입해야하는 구조라서 좋은 구조는 아닌거 같아요,
만약 어쩔수 없는 상황이라면 DefaultValue를 활용해보셔도 좋을거 같아요

namjackson reviewed (changes requested):
안녕하세요 현준님! 
5단계 고생하셨습니다! 
몇가지 추가적인 코멘트 남겼으니, 확인해주세요 😄

jdalma committed:
refactor: 이름 길이 검증 메서드 이름 수정

jdalma committed:
refactor: RacingController, RacingRule 책임 수정 및

1. RacingRule의 이름 길이 검증을 RacingController의 init 블록에서 검증하도록 수정함
2. RacingController의 자동차 이름과 시도 횟수를 RacingPlayers로 래핑

jdalma committed:
refactor: RacingResult 테스트 가능하도록 수정함

1. 경기 결과를 RacingResult 생성시 주입받도록 수정함
2. 경기 우승자 RacingWinner 추가
3. RacingResult 테스트 추가

namjackson commented on src/main/kotlin/racingcar/Main.kt:
```diff
@@ -1,21 +1,31 @@
-    ).gameStart()
+    val nameOfCars = View.printAndGetLine(Message.NAME_OF_CARS_INPUT, Message.INPUT_EXCEPTION)
+    val numberOfTrials = View.printAndGetLineToInt(Message.NUMBER_OF_TRIALS_INPUT, Message.INPUT_NUMBER_EXCEPTION)
+    val players = RacingPlayers(nameOfCars, numberOfTrials)
+
+    val isNotNameLengthOver = { names: List<String> -> names.all { it.length <= NameValidation.NAME_MAX_LENGTH } }
```
RacingPlayers라는 객체는 어떤 역할이 있을까요? 
nameOfCars를 받아서, 해당 검증까지 진행해주는 책임을 가진 객체로 만들어보면 어떨까요?

namjackson commented on src/main/kotlin/racingcar/domain/RacingRule.kt:
```diff
@@ -0,0 +1,8 @@
+package racingcar.domain
+
+class RacingRule(
+    private val random: NumberGenerator<Int>,
+    private val property: RacingProperty
+) {
+    fun canMoveForward(): Boolean = property.forwardCondition <= random.apply(property.start, property.end)
```
고민만 해주셔도 좋아요!
요구사항은 랜덤숫자를 이용한 요구사항이지만,
다른 요구사항 ( ex> 항상 전진하는, 문자열을 받으면 전진하는 )이 추가될수 있을거 같아요, 
그렇다면, RacingRule의 canMoveForward 함수만  추상화해도 좋을거같아요 

```
interface RacingRule { 
 fun canMoveForward(): Boolean 
}

class RandomRacingRule {
 override fun canMoveForward(): Boolean { ...
```

namjackson commented:
마지막 미션이니, 자유롭게 리팩터링해보셔도 좋을거같아요 😄

namjackson reviewed (changes requested):
안녕하세요 현준님, 
몇가지 코멘트 남겼으니, 확인해주세요! 
마지막 미션이니, 자유롭게 리팩터링해보아요!

namjackson commented:
아마 다음 리뷰요청시에는 
미션이 종료처리될거 같아요!
마지막 미션이니, 코드정리나 리팩터링을 해주셔도 좋을거같아요!

jdalma committed:
refactor: RacingPlayers, NameValidation 수정함

1. 이름만 검증하던 NameValidation을 RacingPlayers 자체로 검증할 수 있도록 시그니처 수정함
2. RacingController에서 하던 검증을 RacingPlayers 객체 생성자에서 검증하도록 수정함

jdalma committed:
refactor: RacingController 삭제, RacingProperty 수정 및

1. RacingRule 추상화
2. RacingPlayersTest 추가

jdalma commented:
RacingPlayers는 data class를 사용해야할지 class를 사용해야 할지 고민했습니다
equals와 hashcode가 필요한 상황도 아니고 init 블록이 추가되면서 class로 변경하긴 했는데 리뷰어님은 어떤 기준이 있는지 궁금합니다 ㅎㅎ

namjackson commented on src/main/kotlin/racingcar/domain/RacingPlayers.kt:
```diff
@@ -0,0 +1,19 @@
+
+import racingcar.util.Message
+
+class RacingPlayers(
```
> RacingPlayers는 data class를 사용해야할지 class를 사용해야 할지 고민했습니다
> equals와 hashcode가 필요한 상황도 아니고 init 블록이 추가되면서 class로 변경하긴 했는데 리뷰어님은 어떤 기준이 있는지 궁금합니다 ㅎㅎ

data class를 활용하지 않는다면 사용하지 않는게 좋다고 생각합니다 😄 
저의 기준 또한 equals, hashCode, copy등의 기능이 필요한경우 사용하곤 합니다.
함수형내에서 불변객체를 정의하기 위해서 많이 사용하곤 하는거 같아요! 
다음 미션들에서는 불변성을 잘 활용해보셔도 좋을거같아요 😄 

추가적으로 관련해서 개인적인 생각에 대한 코멘트 드린다면, 
RacingPlayersValidation 객체의 책임은 무엇일까요? 
RacingPlayers내에서 처리되어도 괜찮을거같아요!

또 init시점에 nameOfCars를 names로 변환하는것도 좋지만,
아래처럼 가짜 생성자를 통한 방법도 활용할수 있을거같아요 😮 

```
class RacingPlayers(
    val names: List<String>,,,,
) {
   companion object { 
        fun create(nameOfCars: String): RacingPlayers { ...
   }
```

namjackson commented on src/test/kotlin/racingcar/domain/RacingStadiumTest.kt:
```diff
@@ -0,0 +1,47 @@
+                histories.size shouldBe 3
+
+                histories.forEach { racingHistories ->
+                    racingHistories.all { CAR_RACING_MOVING_SYMBOL.message.repeat(it.move) == CAR_RACING_MOVING_SYMBOL.message.repeat(0) } shouldBe true
```
굳이 CAR_RACING_MOVING_SYMBOL 를 확인하여 검증할 필요가 있을까요?
해당 symbol은 View관련 데이터아닐까요?
`it.move shouldBe 0` 데이터 자체만 검증하면 좋을거같아요!

namjackson commented on src/main/kotlin/racingcar/domain/RacingPlayers.kt:
```diff
@@ -0,0 +1,19 @@
+
+class RacingPlayers(
+    nameOfCars: String,
+    val numberOfTrials: Int,
```
numberOfTrials은 RacingPlayers내에서 관리되어야할 이유가 있을까요?
RacingPlayers 라는 책임과는 어울리지 않을거 같아요! 

`RacingStadium(...).gameStart(numberOfTrials)`
게임 시작시에 파라미터로 전달하는게 좀더 직관적이진 않을까요?

namjackson commented on src/main/kotlin/racingcar/Main.kt:
```diff
@@ -1,21 +1,35 @@
-    ).gameStart()
+    val nameOfCars = View.printAndGetLine(Message.NAME_OF_CARS_INPUT, Message.INPUT_EXCEPTION)
+    val numberOfTrials = View.printAndGetLineToInt(Message.NUMBER_OF_TRIALS_INPUT, Message.INPUT_NUMBER_EXCEPTION)
+    val nameLengthIsLessThan = RacingPlayersValidation { it -> it.names.all { RacingPlayersValidation.NAME_MAX_LENGTH > it.length } }
```
main함수가 조금 복잡해진거같아요,
intRandomGenerator, property, randomRacingRule 등
필요 입력 요소가 아니라면 defaultValue를 통해 정의해두어도 좋을거가 같다는 생각이 들어요 😄

namjackson reviewed (changes requested):
안녕하세요 현준님! 
피드백 적용해주셨네요!
코멘트 주신부분에 대한 부분 다시 코멘트로 남겼습니당 😄

namjackson reviewed (changes requested):
안녕하세요 현준님! 
피드백 적용해주셨네요!
코멘트 주신부분에 대한 부분 다시 코멘트로 남겼습니당 😄

jdalma committed:
refactor: RacingPlayers 책임 수정 및 RacingStadium 수정함

1. RacingPlayers가 가지던 시도 횟수는 게임 시작 파라미터로 수정함
2. RacingStadium이 가지던 cars 내부 프로퍼티를 제거하고 게임 시작시 새로운 자동차를 매번 만들도록 수정함
3. RacingPlayers의 of 메서드 추가

jdalma committed:
refactor: Main 수정 RandomRacingRule 기본 값 선언

jdalma commented:
RacingPlayersValidation은 RacingPlayers 객체에 대한 검증이 추가된다면 해당 인터페이스를 구현하도록 하고 싶었습니다.
지금은 이름 길이에 대한 검증이지만 시도 횟수에 대한 검증이 추가된다면 아래와 같이 사용할 수 있을 것 같았습니다.

```kotlin
class RacingPlayers(
    nameOfCars: String,
    val numberOfTrials: Int,
    val validations: List<RacingPlayersValidation>
)
```

하지만 지금은 이름 길이에 대한 검증만 필요하니 이른 추상화로 볼 수 있을 것 같네요 ㅎㅎ
시도 횟수도 RacingPlayers가 가지고 있어야 할 책임이라고 생각했는데 말씀하신대로 선수와 게임 횟수의 책임을 더 분리할 수 있네요
이런 책임의 단위를 생각하는 것이 어려운 것 같습니다
어떤 객체에 대한 책임의 오류를 스스로 발견하기가 힘드네요

그리고 개발하면서 추상화 지점이 보인다고해서 바로 추상화를 적용할지 고민이 되었습니다.
추상화를 미리 해 놓으면 후에 작업하는 분이 제 의도를 잘 알아차려줄 수 있지 않을까 싶기도 한데, 추상화를 잘 못 해놓으면 다른 작업자가 이해하기가 더 힘들 수도 있을 것 같기도해서.. 고민이 되네요
리뷰어님은 어떤 기준으로 추상화를 적용해야 할 때라고 판단하시는지 궁금합니다!

namjackson commented:
> 시도 횟수도 RacingPlayers가 가지고 있어야 할 책임이라고 생각했는데 말씀하신대로 선수와 게임 횟수의 책임을 더 분리할 수 있네요
> 이런 책임의 단위를 생각하는 것이 어려운 것 같습니다
> 어떤 객체에 대한 책임의 오류를 스스로 발견하기가 힘드네요

저는 개인적으로 객체의 책임에 대해 고민하는편 입니다,
해당 객체의 책임은 무엇인지, 알맞은 네이밍을 가지고 있는지, 이름만으로 어떤 역할을 하는지 예측할수 있는지 
등의 여러 고민을 해보는것도 좋을거같아요
 
> 그리고 개발하면서 추상화 지점이 보인다고해서 바로 추상화를 적용할지 고민이 되었습니다.
> 추상화를 미리 해 놓으면 후에 작업하는 분이 제 의도를 잘 알아차려줄 수 있지 않을까 싶기도 한데, 추상화를 잘 못 해놓으면 다른 작업자가 이해하기가 더 힘들 수도 있을 것 같기도해서.. 고민이 되네요
> 리뷰어님은 어떤 기준으로 추상화를 적용해야 할 때라고 판단하시는지 궁금합니다!

정답이 없는 부분이니 참고만해주셔도 좋을거같아요, 
클린 코드에서 관련된 원칙을 소개해주긴합니다! 
"당신은 그것이 필요하지 않을 것이다"라는 의미의 "You Aren't Gonna Need It" 약자인 YAGNI입니다.
개발자들에게 현재 필요하지 않은 기능을 구현하거나 미래의 요구사항을 미리 예측해서 구현하기보다는,
현재의 요구사항을 충족하고 코드를 간결하게 유지하도록 장합니다! 

자동차미션에서는 테스트 가능한 코드를 위해 랜덤영역을 분리하여 테스트 위해, 필요했었다고 생각이 들어요,
현준님이 `추상화 지점` 생각한 이유가 무엇인지를 고민해보셔도 좋을거 같아요!

추상화를 통해 의미있는 의존성이 분리되었는지, 테스트가 가능해졌는지, 코드나 가독성이 좋아졌는지 등의 고민을 하면서 
현준님만의 기준을 찾아보셔도 좋을거 같습니다!

namjackson reviewed (approved):
안녕하세요 현준님,
자동차 경주 미션 고생하셨습니다! 
미션은 여기서 종료하도록할게요!  😄 
미션하는 과정이 의미있는 시간이 되셨길 바랍니다!
다음 미션들도 응원할게요! 꼭 완주해보아요!
